OSDOCS-11731: Developer perspective for Network Observability (#81044)
openshift/openshift-docs: modules/network-observability-multitenancy.adoc

// network_observability/installing-operators.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-multi-tenancy_{context}"]
= Enabling multi-tenancy in Network Observability
Multi-tenancy in the Network Observability Operator allows and restricts individual user access, or group access, to the flows stored in Loki. Access is enabled for project administrators. Project administrators who have limited access to some namespaces can access flows for only those namespaces. 

For Developers, multi-tenancy is available for both Loki and Prometheus but requires different access rights.

.Prerequisite
* You have installed at least link:https://catalog.redhat.com/software/containers/openshift-logging/loki-rhel8-operator/622b46bcae289285d6fcda39[Loki Operator version 5.7]
* You must be logged in as a project administrator

.Procedure

> * To enable limited access to some namespaces for the administrator perspective, authorize reading permission to the `<user_group_or_name>` by running the following command:

Reviewer: The role `cluster-monitoring-view` does actually give access to cluster-wide metrics, so it is not "limited to some namespaces". I think this would be how it goes:
1. For per-tenant access, users should have the `netobserv-reader` cluster role and the `netobserv-metrics-reader` namespaced role, and use preferably the developer perspective. The admin perspective would work only with Loki.
2. For cluster wide access, users (who are not cluster admins) should have the `netobserv-reader`  and the `cluster-monitoring-view` cluster roles. They can use either the admin perspective or the developer perspective, as they wish.

cc @Amoghrd / @memodi does it sound good to you?

Reviewer: (Note that in the second case, the command "oc adm policy add-cluster-role-to-user netobserv-metrics-reader <user_group_or_name>" is only needed until [this bug](https://issues.redhat.com/browse/OCPBUGS-41158) is fixed on the cluster monitoring side)

Author: Thanks for the details @jotak . I've updated the documentation accordingly.

// network_observability/installing-operators.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-multi-tenancy_{context}"]
= Enabling multi-tenancy in Network Observability

> Multi-tenancy in the Network Observability Operator allows and restricts individual user access, or group access, to the flows stored in Loki. Access is enabled for project administrators. Project administrators who have limited access to some namespaces can access flows for only those namespaces. 

Author: Just want to put this here because its going to look weird, but according to IBM style guide, I have to use "and or or" and not "and/or" per but I will take this suggestion, as long as the project admin stuff isn't important anymore. I agree I should say Prometheus here too. 
> Do not use the construction and/or, which can be ambiguous and cause problems for translation. Instead, consider whether both conjunctions are needed; if not, use and or or. If both conjunctions are needed, use the construction option A, option B, or both.

For Developers, multi-tenancy is available for both Loki and Prometheus but requires different access rights.

.Prerequisite

> * You have installed at least link:https://catalog.redhat.com/software/containers/openshift-logging/loki-rhel8-operator/622b46bcae289285d6fcda39[Loki Operator version 5.7]

Reviewer 2: Does the prereq to install Loki is still needed? If they are using prom mode then they dont need to install Loki

Author: Hmmmm maybe it needs to be updated. To "If you are using Loki, you have installed at least ..."

* You must be logged in as a project administrator

.Procedure

> * For per-tenant access, you must have the `netobserv-reader` cluster role and the `netobserv-metrics-reader` namespace role. In this scenario, it is preferable to use the developer perspective. The administrator perspective only works with Loki. Run the following commands for this level of access:

Reviewer 3: ```suggestion
* For per-tenant access, use the developer perspective. The administrator perspective only works with Loki. Run the following commands to get the per-tenant level of access:
```
- This can be updated to say something like -- 
"for this level of access" -- "to get the per-tenant level of access" 
- This  info "For per-tenant access, you must have the `netobserv-reader` cluster role and the `netobserv-metrics-reader` namespace role" can be removed from here and be used as instructions for the later commands. For example - "Add the netobserv-reader cluster role to.....by running the following..."

> * For per-tenant access, you must have the `netobserv-reader` cluster role and the `netobserv-metrics-reader` namespace role. In this scenario, it is preferable to use the developer perspective. The administrator perspective only works with Loki. Run the following commands for this level of access:

Reviewer 4: ```suggestion
* For per-tenant access, you must have the `netobserv-reader` cluster role and the `netobserv-metrics-reader` namespace role to use the developer perspective. Run the following commands for this level of access:
```

> +

Reviewer 3: Consider adding an instruction for this command: 
"Add the `netobserv-reader` cluster role to.....by running the following..."

This can be a substep.

[source,terminal]
----
$ oc adm policy add-cluster-role-to-user netobserv-reader <user_group_or_name>
----

> +

Reviewer 3: Consider adding an instruction for this command: 

"Add the `netobserv-metrics-reader` namespace role to.....by running the following..."

This can be a substep.

[source,terminal]
----
$ oc adm policy add-role-to-user netobserv-metrics-reader <user_group_or_name> -n <namespace>
----

* For cluster-wide access, non-cluster-administrators must have the `netobserv-reader`, `cluster-monitoring-view`, and `netobserv-metrics-reader` cluster roles. In this scenario, you can use either the admin perspective or the developer perspective. Run the following commands for this level of access:
+
[source,terminal]
----
$ oc adm policy add-cluster-role-to-user netobserv-reader <user_group_or_name>

> ----

Reviewer 3: Same comments as earlier

+
[source,terminal]
----

> $ oc adm policy add-cluster-role-to-user cluster-monitoring-view <user_group_or_name>

Reviewer 3: Same comment about adding an instruction for this command.

:_mod-docs-content-type: PROCEDURE
[id="network-observability-multi-tenancy{context}"]
= Enabling multi-tenancy in Network Observability
Multi-tenancy in the Network Observability Operator allows and restricts individual user access, or group access, to the flows stored in Loki. Access is enabled for project administrators. Project administrators who have limited access to some namespaces can access flows for only those namespaces. 

For Developers, multi-tenancy is available for both Loki and Prometheus but requires different access rights.

.Prerequisite
* You have installed at least link:https://catalog.redhat.com/software/containers/openshift-logging/loki-rhel8-operator/622b46bcae289285d6fcda39[Loki Operator version 5.7]

> * You must be logged in as a project administrator or developer

Reviewer: actually, they must be logged in as cluster admins. It's admins who can grant permissions to less privileged users

+
Now, the data is restricted to only allowed user namespaces. For example, a user that has access to a single namespace can see all the flows internal to this namespace, as well as flows going from and to this namespace.
Project admins have access to the Administrator perspective in the {product-title} console to access the Network Flows Traffic page.

* To enable multi-tenancy for developer perspective, run the following command:
+
[source,terminal]
----

> $ oc adm policy add-role-to-user netobserv-metrics-reader <user_group_or_name> -n <your_namespace>

Reviewer: in the above command we use an example user name `user1` and here we use a placeholder `<user_group_or_name>`, I guess we should align the two commands in a way or another

Author: Good point I will make them similar, user-replaced values.

> ----

Author: @jotak @jpinsonneau Ok, how does this look? I'm unsure if I should add it here or not, since technically this section is nested under the Loki stuff. I could bring it out from underneath the Loki stuff though, since it applies to non-Loki use cases as well. 

Is there anything else we need to say besides MT is available for Loki and Prometheus in Developer perspective but needs this access? If this is OK, I will link it to the Developer perspective info in the Overview too.

Reviewer 4: Can't we get this section right after `Installing the Network Observability Operator`, at the same level ?

It will make more sense as we need to install the operator to get the roles created.
Then we can have 2 optional cases for Loki and Prometheus.

Reviewer: +1 @jpinsonneau 
By the way, we should also mention the `cluster-monitoring-view` role I guess, for the admin perspective? Users might very well already have it because it's not specific to netobserv, but a project admin who uses the administrator perspective needs that role too:

```bash
$ oc adm policy add-cluster-role-to-user cluster-monitoring-view <user_group_or_name>
```

Author: @jotak so for the admin perspective, the command would be this?:
`$ oc adm policy add-cluster-role-to-user cluster-monitoring-view netobserv-reader <user_group_or_name>`

Or should I separate the `cluster-monitoring-view`?
